sdk/keyvault/keyvault-common/src/keyVaultAuthenticationPolicy.ts

```diff
@@ -156,25 +173,105 @@ export function createKeyVaultChallengeCallbacks(
 
     const accessToken = await getAccessToken([scope], {
       ...getTokenOptions,
+      enableCae: true,
       tenantId: parsedChallenge.tenantId,
     });
 
     if (!accessToken) {
-      return false;
+      // No access token provided, treat as no-op
+      return response;
     }
 
     request.headers.set("Authorization", `Bearer ${accessToken.token}`);
 
     challengeState = {
       status: "complete",
       scopes: [scope],
+      tenantId: parsedChallenge.tenantId,
     };
 
-    return true;
+    // We have a token now, so try send the request again
+    return next(request);
+  }
+
+  async function handleCaeChallenge(
+    request: PipelineRequest,
+    response: PipelineResponse,
+    next: SendRequest,
+  ): Promise<PipelineResponse> {
+    // Cannot handle CAE challenge if a regular challenge has not been completed first
+    if (challengeState.status !== "complete") {
+      return response;
+    }
+
+    // If status is not 401, this is a no-op
+    if (response.status !== 401) {
+      return response;
+    }
+
+    const getTokenOptions = requestToOptions(request);
+
+    const challenge = response.headers.get("WWW-Authenticate");
+    if (!challenge) {
+      throw new Error(
+        "A 401 response from the service must have a WWW-Authenticate header. This is a service bug.",
+      );
+    }
+    const {
+      claims: base64EncodedClaims,
+      error,
+      scope: newScope,
+      tenantId: newTenantId,
+    }: WWWAuthenticate = parseWWWAuthenticateHeader(challenge) || {};
+
+    if (error !== "insufficient_claims" || base64EncodedClaims === undefined) {
+      return response;
+    }
+
+    const claims = atob(base64EncodedClaims);
+
+    const scopes = newScope ? [newScope] : challengeState.scopes;
+    const tenantId = newTenantId ?? challengeState.tenantId;
```

Comment:
You can delete this. CAE is driven by conditional access policies defined for individual tenants. A CAE challenge implies the presented token was from the correct tenant because otherwise Key Vault would have responded with its own challenge before evaluating conditional access policies. So for example, when a vault in tenant A receives a token from tenant B, it doesn't evaluate any conditional access policies because A's policies don't apply to B's token and B's policies are irrelevant because even if the token satisfies all of them, the vault would still reject it as being from the wrong tenant. The vault simply responds with an ordinary challenge telling the client to come back with a token from A

Reply:
Removed